// @tanstack/angular-query-experimental/build/esm2022/util/create-injection-token/create-injection-token.mjs
/* eslint-disable cspell/spellchecker */
/**
 * The code in this file is adapted from NG Extension Platform at https://ngxtension.netlify.app.
 *
 * Original Author: Chau Tran
 *
 * NG Extension Platform is an open-source project licensed under the MIT license.
 *
 * For more information about the original code, see
 * https://github.com/nartc/ngxtension-platform
 */
/* eslint-enable */
import { InjectionToken, inject, runInInjectionContext } from '@angular/core';
import { assertInjector } from '../assert-injector/assert-injector';
function createInjectFn(token) {
    return function ({ injector, ...injectOptions } = {}) {
        injector = assertInjector(this, injector);
        return runInInjectionContext(injector, () => inject(token, injectOptions));
    };
}
function createProvideFn(token, factory, opts = {}) {
    const { deps = [], multi = false, extraProviders = [] } = opts;
    return (value, isFunctionValue = false) => {
        let provider;
        if (value !== undefined) {
            // eslint-disable-next-line no-shadow
            const factory = typeof value === 'function'
                ? isFunctionValue
                    ? () => value
                    : value
                : () => value;
            provider = {
                provide: token,
                useFactory: factory,
                multi,
            };
        }
        else {
            provider = {
                provide: token,
                useFactory: factory,
                deps: deps,
                multi,
            };
        }
        return [extraProviders, provider];
    };
}
export function createNoopInjectionToken(description, options) {
    const token = 
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    options?.token ||
        new InjectionToken(description);
    return [
        createInjectFn(token),
        createProvideFn(token, () => null, (options || {})),
        token,
        () => { },
    ];
}
